eth/downloader/queue.go

```diff
@@ -370,81 +571,155 @@ func (q *queue) Reserve(p *peer, count int) (*fetchRequest, bool, error) {
 	}
 	// Merge all the skipped headers back
 	for _, header := range skip {
-		q.headerQueue.Push(header, -float32(header.Number.Uint64()))
+		taskQueue.Push(header, -float32(header.Number.Uint64()))
 	}
 	// Assemble and return the block download request
 	if len(send) == 0 {
-		return nil, process, nil
+		return nil, progress, nil
 	}
 	request := &fetchRequest{
 		Peer:    p,
 		Headers: send,
 		Time:    time.Now(),
 	}
-	q.pendPool[p.id] = request
+	pendPool[p.id] = request
+
+	return request, progress, nil
+}
+
+// CancelBlocks aborts a fetch request, returning all pending hashes to the queue.
+func (q *queue) CancelBlocks(request *fetchRequest) {
+	q.cancel(request, q.hashQueue, q.blockPendPool)
+}
+
+// CancelBodies aborts a body fetch request, returning all pending headers to the
+// task queue.
+func (q *queue) CancelBodies(request *fetchRequest) {
+	q.cancel(request, q.blockTaskQueue, q.blockPendPool)
+}
 
-	return request, process, nil
+// CancelReceipts aborts a body fetch request, returning all pending headers to
+// the task queue.
+func (q *queue) CancelReceipts(request *fetchRequest) {
+	q.cancel(request, q.receiptTaskQueue, q.receiptPendPool)
 }
 
-// Cancel aborts a fetch request, returning all pending hashes to the queue.
-func (q *queue) Cancel(request *fetchRequest) {
+// CancelNodeData aborts a node state data fetch request, returning all pending
+// hashes to the task queue.
+func (q *queue) CancelNodeData(request *fetchRequest) {
+	q.cancel(request, q.stateTaskQueue, q.statePendPool)
+}
+
+// Cancel aborts a fetch request, returning all pending hashes to the task queue.
+func (q *queue) cancel(request *fetchRequest, taskQueue *prque.Prque, pendPool map[string]*fetchRequest) {
 	q.lock.Lock()
 	defer q.lock.Unlock()
 
 	for hash, index := range request.Hashes {
-		q.hashQueue.Push(hash, float32(index))
+		taskQueue.Push(hash, float32(index))
 	}
 	for _, header := range request.Headers {
-		q.headerQueue.Push(header, -float32(header.Number.Uint64()))
+		taskQueue.Push(header, -float32(header.Number.Uint64()))
+	}
+	delete(pendPool, request.Peer.id)
+}
+
+// Revoke cancels all pending requests belonging to a given peer. This method is
+// meant to be called during a peer drop to quickly reassign owned data fetches
```

Comment:
It means, that whenever the downloader needs to pull some data, it assigns that piece of missing data (its hash) to a peer to avoid requesting the same data from somebody else too. Now, if we drop a peer, we need to return all missing data assigned to it to the pool so that others might fulfill them.